pubsub/pubsub_goredis.go

```diff
@@ -83,88 +85,58 @@ func (ps *GoRedisPubSub) Stop() error {
 	return nil
 }
 
-// assert that GoRedisPubSub implements PubSub
-var _ PubSub = (*GoRedisPubSub)(nil)
-
-// when a topic is created, it is stored in the topics map and a goroutine is
-// started to listen for messages on the topic; each message is sent to all
-// subscribers to the topic.
-func (ps *GoRedisPubSub) NewTopic(ctx context.Context, topic string) Topic {
-	t := &GoRedisTopic{
-		client:      ps.client,
-		topic:       topic,
-		redisSub:    ps.client.Subscribe(ctx, topic),
-		subscribers: make([]chan string, 0),
-		done:        make(chan struct{}),
+func (ps *GoRedisPubSub) Close() {
+	ps.mut.Lock()
+	defer ps.mut.Unlock()
+	for _, sub := range ps.subs {
+		sub.Close()
 	}
+	ps.subs = nil
+	ps.client.Close()
+}
+
+func (ps *GoRedisPubSub) Publish(ctx context.Context, topic, message string) error {
+	ps.mut.RLock()
+	defer ps.mut.RUnlock()
+	return ps.client.Publish(ctx, topic, message).Err()
+}
 
+func (ps *GoRedisPubSub) Subscribe(ctx context.Context, topic string) Subscription {
+	ps.mut.Lock()
```

Comment:
We're accessing ps.client when we build the Subscription object. 

I could move the unlock to above the goroutine, but I don't think it makes a difference since the only other thing that happens is the return.

Reply:
I believe the redis.Client is concurrent safe based on their doc